core/sourcehook/sourcehook_hookmangen_x86_64.cpp

```diff
@@ -401,7 +404,17 @@ namespace SourceHook
 				}
 			}
 #else
-static_assert(false, "Missing registers saving for linux");
+			const x86_64_Reg params_reg[] = { rdi, rsi, rdx, rcx, r8, r9 };
+
+			int reg_index = 0;
+		
+			m_HookFunc.mov(rbp(v_this), rdi);
+			reg_index++;
+
+			if ((retInfo.flags & PassInfo::PassFlag_RetMem) == PassInfo::PassFlag_RetMem) {
+				m_HookFunc.mov(rbp(v_memret_ptr), params_reg[reg_index]);
+				reg_index++;
+			}
```

Comment:
Given that System V callconv is very annoying because sourcehook doesn't provide nearly enough information to tell whether an object goes in memory or registers. I'm fine with the idea of just saving all the registers and restoring them when its time to call the original. Only the return value will therefore need special attention.

It's very important all the registers are saved, because in the rest of the stub generator I make the promise that all general purpose registers are available for usage. Which I did make use of (see when `r8` or `r9` get used) and they not restroed, this will create issues.

Reply:
> Given that System V callconv is very annoying because sourcehook doesn't provide nearly enough information to tell whether an object goes in memory or registers

This is nice in a way though because it means all the type classification logic from the ABI can be excluded.

Which does mean Linux64 Sourcepawn DHooks users will have to manually fill out registers, where on Windows64 you can just `HookParamType_Object`/`HookParamType_ObjectPtr` and call it a day...